README.md
<a href="http://www.bioconductor.org/packages/release/bioc/html/regionReport.html#since"><img border="0" src="http://www.bioconductor.org/shields/years-in-bioc/regionReport.svg" title="How long since the package was first in a released Bioconductor version (or is it in devel only)."></a> <a href="https://bioconductor.org/packages/stats/bioc/regionReport/"><img border="0" src="http://www.bioconductor.org/shields/downloads/regionReport.svg" title="Percentile (top 5/20/50% or 'available') of downloads over last 6 full months. Comparison is done across all package categories (software, annotation, experiment)."></a> <a href="https://support.bioconductor.org/t/regionReport/"><img border="0" src="http://www.bioconductor.org/shields/posts/regionReport.svg" title="Support site activity, last 6 months: tagged questions/avg. answers per question/avg. comments per question/accepted answers, or 0 if no tagged posts."></a> <a href="http://www.bioconductor.org/packages/release/bioc/html/regionReport.html#svn_source"><img border="0" src="http://www.bioconductor.org/shields/commits/bioc/regionReport.svg" title="average Subversion commits (to the devel branch) per month for the last 6 months"></a>

Status: Travis CI [![Build Status](https://travis-ci.org/leekgroup/regionReport.svg?branch=master)](https://travis-ci.org/leekgroup/regionReport),
Bioc-release <a href="http://www.bioconductor.org/packages/release/bioc/html/regionReport.html#archives"><img border="0" src="http://www.bioconductor.org/shields/availability/release/regionReport.svg" title="Whether the package is available on all platforms; click for details."></a> <a href="http://bioconductor.org/checkResults/release/bioc-LATEST/regionReport/"><img border="0" src="http://www.bioconductor.org/shields/build/release/bioc/regionReport.svg" title="build results; click for full report"></a>,
Bioc-devel <a href="http://www.bioconductor.org/packages/devel/bioc/html/regionReport.html#archives"><img border="0" src="http://www.bioconductor.org/shields/availability/devel/regionReport.svg" title="Whether the package is available on all platforms; click for details."></a> <a href="http://bioconductor.org/checkResults/devel/bioc-LATEST/regionReport/"><img border="0" src="http://www.bioconductor.org/shields/build/devel/bioc/regionReport.svg" title="build results; click for full report"></a>.

Bioc-release <a href="https://bioconductor.org/developers/how-to/unitTesting-guidelines/#coverage"><img border="0" src="http://www.bioconductor.org/shields/coverage/release/regionReport.svg" title="Test coverage percentage, or 'unknown'"></a>, Bioc-devel <a href="https://bioconductor.org/developers/how-to/unitTesting-guidelines/#coverage"><img border="0" src="http://www.bioconductor.org/shields/coverage/devel/regionReport.svg" title="Test coverage percentage, or 'unknown'"></a>

regionReport
===============

Generate HTML reports for a set of regions such as those from 
[derfinder](http://www.bioconductor.org/packages/derfinder) results or any other pipeline that defines a set of genomic regions.

Check the documentation for _derfinderReport()_ for an example on how to create 
the necessary input files and generating the HTML report for __derfinder__ 
results. Or use:

```R
example('derfinderReport', 'regionReport', ask=FALSE)
```

Similarly, check _renderReport()_ for an example of a general report, or use:

```R
example('renderReport', 'regionReport', ask=FALSE)
```

For DESeq2 or edgeR results check _DESeq2Report()_ and _edgeReport()_.


You may also want to consult the [vignette](http://www.bioconductor.org/packages/regionReport) for `regionReport` as well as the supplementary website [regionReportSupp](http://leekgroup.github.io/regionReportSupp/).

# Installation instructions

Get R 3.5.x from [CRAN](http://cran.r-project.org/) as well as pandoc 
1.17.0.3 or newer from [here](http://johnmacfarlane.net/pandoc/installing.html).

```R
## From Bioconductor
install.packages("BiocManager")
BiocManager::install('regionReport')
```

# Vignette

The vignette for this package can be viewed via [Bioconductor's website](http://www.bioconductor.org/packages/regionReport).


# Citation

Below is the citation output from using `citation('regionReport')` in R. 
Please run this yourself to check for any updates on how to cite 
__regionReport__.


To cite package __regionReport__ in publications use:

Collado-Torres L, Jaffe AE and Leek JT (2016). “regionReport: Interactive reports for region-level and feature-level genomic analyses [version2; referees: 2 approved, 1 approved with reservations].” _F1000Research_, *4*, pp. 105. doi: 10.12688/f1000research.6379.2 (URL:
http://doi.org/10.12688/f1000research.6379.2), <URL: http://f1000research.com/articles/4-105/v2>.


A BibTeX entry for LaTeX users is

@Article{,
    author = {Leonardo Collado-Torres and Andrew E. Jaffe and Jeffrey T. Leek},
    title = {regionReport: Interactive reports for region-level and feature-level genomic analyses [version2; referees: 2 approved, 1 approved with reservations]},
    journal = {F1000Research},
    year = {2016},
    doi = {10.12688/f1000research.6379.2},
    url = {http://f1000research.com/articles/4-105/v2},
    volume = {4},
    pages = {105},
}

# Testing

Testing on Bioc-devel is feasible thanks to [R Travis](http://docs.travis-ci.com/user/languages/r/) as well as Bioconductor's nightly build.


NEWS
CHANGES IN VERSION 1.18.2
-------------------------

SIGNIFICANT USER-VISIBLE CHANGES

    o Renamed load_install() as load_check() as this function now only checks
        that the package(s) was installed and returns an error if missing. The
        error shows the user how to install the package(s) they are missing
        instead of installing them automatically. This complies with Marcel
        Ramos' request at https://github.com/leekgroup/recount/issues/14.


CHANGES IN VERSION 1.17.2
-------------------------

NEW FEATURES

    o Add ORCID's following changes at
        http://bioconductor.org/developers/package-guidelines/#description

CHANGES IN VERSION 1.15.4
-------------------------

BUG FIXES

    o Fix a bug in the order that was reported and fixed by @bounlu at
        https://github.com/leekgroup/regionReport/pull/9/


CHANGES IN VERSION 1.15.3
-------------------------

BUG FIXES

    o Fixed an issue with DESeq2Exploration.Rmd that affected both
        DESeq2Report and edgeReport. This should also fix the recount
        bioc-release (3.7) and bioc-devel (3.8) branches.

    o Fixed a NAMESPACE issue with rmarkdown::html_document and
        BiocStyle::html_document


CHANGES IN VERSION 1.15.2
-------------------------

SIGNIFICANT USER-VISIBLE CHANGES

    o Use BiocManager

CHANGES IN VERSION 1.15.1
-------------------------

BUG FIXES

    o Fix namespace issue in relation to BiocStyle::html_document2

CHANGES IN VERSION 1.11.6
-------------------------

SIGNIFICANT USER-VISIBLE CHANGES

    o Changed the default style to BiocStyle::html_document to mirror recent
        changes in BiocStyle.


CHANGES IN VERSION 1.11.4
-------------------------

SIGNIFICANT USER-VISIBLE CHANGES

    o Vignette now uses the new BiocStyle::html_document that was recently
        released.


CHANGES IN VERSION 1.11.2
-------------------------

BUG FIXES

    o Fixed the citations.
    o Fixed DESeq2Report() for limma-results so that it will properly cite
        limma.

CHANGES IN VERSION 1.11.1
-------------------------

SIGNIFICANT USER-VISIBLE CHANGES

    o DESeq2Report() can now be used with other software if their results are
         made to look like DESeq2 results. For example, with limma-voom results.

BUG FIXES

    o Made the DESeq2Report() more robust in case rlog() fails initially.


CHANGES IN VERSION 1.9.1
------------------------

SIGNIFICANT USER-VISIBLE CHANGES

    o Changed the default style to BiocStyle::html_document2.

CHANGES IN VERSION 1.7.10
-------------------------

SIGNIFICANT USER-VISIBLE CHANGES

    o Help pages now document advanced arguments.

CHANGES IN VERSION 1.7.2
------------------------

SIGNIFICANT USER-VISIBLE CHANGES

    o Dropped defunct functions.


CHANGES IN VERSION 1.5.48
-------------------------

BUG FIXES

    o Fixed a bug in derfinderReport() for a case when there are significant
        regions but not all regions have finite areas.

CHANGES IN VERSION 1.5.43
-------------------------

SIGNIFICANT USER-VISIBLE CHANGES

    o edgeReport() now includes two edgeR specific plots: one showing the BCV
        and another showing a 2-dim MDS. Also added more edgeR citations that I
        missed earlier: thank you Gordon Smyth!

CHANGES IN VERSION 1.5.33
-------------------------

NEW FEATURES

    o Added the function edgeReport() for creating HTML or PDF reports based on
        edgeR results. Together with DESeq2Report() now regionReport supports
        the two most used packages for RNA-seq feature-level analysis.


CHANGES IN VERSION 1.5.19
-------------------------

NEW FEATURES

    o Added the templates 'templatePvalueHistogram' and 'templateHistogram'
        to be used with renderReport() if you prefer histogram plots instead
        of density plots.

CHANGES IN VERSION 1.5.12
-------------------------

NEW FEATURES

    o Added the function DESeq2Report() for creating HTML or PDF reports based
        on DESeq2 results. This should also be useful to explore derfinder
        results created from the expressed regions-level approach.

SIGNIFICANT USER-VISIBLE CHANGES

    o Added a 'digits' argument to control how to round some numerical variables
        in all type of reports.
    o Added a 'theme' argument to allow setting the ggplot2 theme for the plots.

BUG FIXES

    o Improved the PDF versions of all reports by hiding code and shortening
        tables. Also added a warning to switch the device to 'pdf' for PDF
        output since it looks better than the default 'png'.

CHANGES IN VERSION 1.5.6
------------------------

SIGNIFICANT USER-VISIBLE CHANGES

    o Switched to using rmarkdown instead of knitrBootstrap as the default
        engine for creating the reports.

CHANGES IN VERSION 1.3.8
------------------------

SIGNIFICANT USER-VISIBLE CHANGES

    o renderReport() and derfinderReport() now show Manhattan plots for p-value
        variables (p-value, q-value, FWER adjusted p-value).


CHANGES IN VERSION 1.3.7
------------------------

NEW FEATURES

    o renderReport() now has the 'densityTemplates' argument via which users can
        customize the density plots for the p-value variables and the continuous
        variables. This addresses one of David Robinson's requests at
        http://f1000research.com/articles/4-105/v1



CHANGES IN VERSION 1.3.6
------------------------

NEW FEATURES

    o Added a vignette with an example report from bumphunter results.

R/renderReport.R
#' Generate a HTML/PDF report exploring a set of genomic regions
#'
#' This function generates a HTML report with quality checks, genome location
#' exploration, and an interactive table with the results. Other output formats
#' are possible such as PDF but lose the interactivity. Users can easily append
#' to the report by providing a R Markdown file to \code{customCode}, or can
#' customize the entire template by providing an R Markdown file to
#' \code{template}.
#'
#' @param regions The set of genomic regions of interest as a \code{GRanges}
#' object. All sequence lengths must be provided.
#' @param project The title of the project.
#' @param pvalueVars The names of the variables with values between 0 and 1 to
#' plot density values by chromosome and a table for commonly used cutoffs.
#' Most commonly used to explore p-value distributions. If a named
#' character vector is provided, the names are used in the plot titles.
#' @param densityVars The names of variables to use for making density plots
#' by chromosome. Commonly used to explore scores and other variables given
#' by region.  If a named character vector is provided, the names are used in
#' the plot titles.
#' @param significantVar A \code{logical} variable differentiating statistically
#' significant regions from the rest. When provided, both types of regions
#' are compared against each other to see differences in width, location, etc.
#' @param annotation The output from \link[bumphunter]{matchGenes} used on
#' \code{regions}. Note that this can take time for a large set of regions
#' so it's better to pre-compute this information and save it.
#' @param nBestRegions The number of regions to include in the interactive
#' table.
#' @param customCode An absolute path to a child R Markdown file with code to be
#' evaluated before the reproducibility section. Its useful for users who want
#' to customize the report by adding conclusions derived from the data and/or
#' further quality checks and plots.
#' @param outdir The name of output directory.
#' @param output The name of output HTML file (without the html extension).
#' @param browse If \code{TRUE} the HTML report is opened in your browser once
#' it's completed.
#' @param txdb Specify the transcription database to use for identifying the
#' closest genes via \link[bumphunter]{matchGenes}. If \code{NULL} it will
#' use TxDb.Hsapiens.UCSC.hg19.knownGene by default.
#' @param device The graphical device used when knitting. See more at
#' http://yihui.name/knitr/options (\code{dev} argument).
#' @param densityTemplates A list of length 3 with templates for the p-value
#' density plots (variables from \code{pvalueVars}), the continuous
#' variables density plots (variables from \code{densityVars}), and Manhattan
#' plots for the p-value variables (\code{pvalueVars}). These templates
#' are processed by \link[whisker]{whisker.render}. Check the default templates
#' for more information. The \code{densityTemplates} argument is available for
#' those users interested in customizing these plots. For example, to show
#' histograms instead of density plots use \code{templatePvalueHistogram} and
#' \code{templateHistogram} instead of \code{templatePvalueDensity} and
#' \code{templateDensity} respectively.
#' @param template Template file to use for the report. If not provided, will
#' use the default file found in regionExploration/regionExploration.Rmd
#' within the package source.
#' @param theme A ggplot2 \link[ggplot2]{theme} to use for the plots made with
#' ggplot2.
#' @param digits The number of digits to round to in the interactive table of
#' the top \code{nBestRegions}. Note that p-values and adjusted p-values won't
#' be rounded.
#' @param ... Arguments passed to other methods and/or advanced arguments.
#' Advanced arguments:
#' \describe{
#' \item{overviewParams }{ A two element list with \code{base_size} and
#' \code{areaRel} that control the text size for the genomic overview plots.}
#' \item{output_format }{ Either \code{html_document}, \code{pdf_document} or
#' \code{knitrBootstrap::bootstrap_document} unless you modify the YAML
#' template.}
#' \item{clean }{ Logical, whether to clean the results or not. Passed to
#' \link[rmarkdown]{render}.}
#' }
#'
#' @return An HTML report with a basic exploration for the given set of
#' genomic regions.
#'
#' @author Leonardo Collado-Torres
#' @export
#'
#' @importFrom knitcitations cleanbib cite_options write.bibtex read.bibtex
#' citep bibliography
#' @importFrom GenomeInfoDb seqlengths
#' @importFrom utils browseURL citation packageVersion
#' @importFrom rmarkdown render
#' @importFrom GenomicRanges mcols 'mcols<-'
#' @importFrom knitrBootstrap knit_bootstrap
#' @importFrom BiocStyle html_document
#' @importFrom methods is
#'
#' @details
#' Set \code{output_format} to \code{'knitrBootstrap::bootstrap_document'} or
#' \code{'pdf_document'} if you want a HTML report styled by knitrBootstrap or
#' a PDF report respectively. If using knitrBootstrap, we recommend the version
#' available only via GitHub at https://github.com/jimhester/knitrBootstrap
#' which has nicer features than the current version available via CRAN. You can
#' also set the \code{output_format} to \code{'html_document'} for a HTML
#' report styled by rmarkdown. The default is set to
#' \code{'BiocStyle::html_document'}.
#'
#' If you modify the YAML front matter of \code{template}, you can use other
#' values for \code{output_format}.
#'
#' The HTML report styled with knitrBootstrap can be smaller in size than the
#' \code{'html_document'} report.
#'
#' @examples
#'
#' ## Load derfinder for an example set of regions
#' library('derfinder')
#' regions <- genomeRegions$regions
#'
#' ## Assign chr length
#' library('GenomicRanges')
#' seqlengths(regions) <- c('chr21' = 48129895)
#'
#' ## The output will be saved in the 'renderReport-example' directory
#' dir.create('renderReport-example', showWarnings = FALSE, recursive = TRUE)
#'
#' ## Generate the HTML report
#' report <- renderReport(regions, 'Example run', pvalueVars = c(
#'     'Q-values' = 'qvalues', 'P-values' = 'pvalues'), densityVars = c(
#'     'Area' = 'area', 'Mean coverage' = 'meanCoverage'),
#'     significantVar = regions$qvalues <= 0.05, nBestRegions = 20,
#'     outdir = 'renderReport-example')
#'
#' if(interactive()) {
#'     ## Browse the report
#'     browseURL(report)
#' }
#'
#' \dontrun{
#' ## Note that you can run the example using:
#' example('renderReport', 'regionReport', ask=FALSE)
#' }
#'
#' ## Check the default templates. For users interested in customizing these
#' ## plots.
#' ## For p-value variables:
#' cat(templatePvalueDensity)
#'
#' ## For continous variables:
#' cat(templateDensity)
#'
#' ## For Manhattan plots
#' cat(templateManhattan)
#'


renderReport <- function(regions, project = "",
    pvalueVars = c('P-values' = 'pval'),
    densityVars = NULL, significantVar = mcols(regions)$pval <= 0.05,
    annotation = NULL, nBestRegions = 500, customCode = NULL,
    outdir = 'regionExploration', output = 'regionExploration',
    browse = interactive(), txdb = NULL, device = 'png',
    densityTemplates = list(Pvalue = templatePvalueDensity,
        Common = templateDensity, Manhattan = templateManhattan),
    template = NULL, theme = NULL, digits = 2, ...) {
    ## Save start time for getting the total processing time
    startTime <- Sys.time()


    ## Check inputs
    stopifnot(is(regions, 'GRanges'))
    stopifnot(!any(is.na(seqlengths(regions))))
    stopifnot(is.list(densityTemplates) & length(densityTemplates) == 3 & all(c('Pvalue', 'Common', 'Manhattan') %in% names(densityTemplates)))
    hasCustomCode <- !is.null(customCode)
    if(hasCustomCode) stopifnot(length(customCode) == 1)
    if(!is.null(annotation)) stopifnot(nrow(annotation) == length(regions))
    if(!is.null(theme)) stopifnot(is(theme, c('theme', 'gg')))

    # @param overviewParams A two element list with \code{base_size} and
    # \code{areaRel} that control the text size for the genomic overview plots.
    overviewParams <- .advanced_argument('overviewParam', list(base_size = 10,
            areaRel = 5), ...)


    ## Check that overviewParams is correctly specified
    stopifnot(sum(names(overviewParams) %in% c('base_size', 'areaRel')) == 2)

    ## Are there p-value vars?
    hasPvalueVars <- length(pvalueVars) > 0
    if(hasPvalueVars) stopifnot(is.character(pvalueVars))
    ## Fix p-value variable names
    colnames(mcols(regions))[which(colnames(mcols(regions)) %in% pvalueVars)] <- make.names(pvalueVars)
    pvalueVars[seq_len(length(pvalueVars))] <- make.names(pvalueVars)

    ## Are there density vars?
    hasDensityVars <- length(densityVars) > 0
    if(hasDensityVars) stopifnot(is.character(densityVars))
    ## Fix density variable names
    colnames(mcols(regions))[which(colnames(mcols(regions)) %in% densityVars)] <- make.names(densityVars)
    densityVars[seq_len(length(densityVars))] <- make.names(densityVars)

    ## Are there significant regions?
    hasSignificant <- length(significantVar) > 0
    if(hasSignificant) {
        stopifnot(length(significantVar) == length(regions))
        stopifnot(is.logical(significantVar))
        hasSignificant <- any(significantVar)
        if(!hasSignificant) warning("There are no statistically significant regions in this set.")
    }

    ## Template for pvalueVars
    if(hasPvalueVars) {
        templatePvalueDensityInUse <- densityTemplates$Pvalue
        templateManhattanInUse <- densityTemplates$Manhattan
    }

    ## Template for densityVars
    if(hasDensityVars) templateDensityInUse <- densityTemplates$Common


    ## Create outdir
    dir.create(outdir, showWarnings = FALSE, recursive = TRUE)
    workingDir <- getwd()

    ## Locate Rmd if one is not provided
    if (is.null(template)) {
        templateNull <- TRUE
        template <- system.file(
            file.path('regionExploration', 'regionExploration.Rmd'),
            package = 'regionReport', mustWork = TRUE
        )
    } else {
        templateNull <- FALSE
    }

    ## Load knitcitations with a clean bibliography
    cleanbib()
    cite_options(hyperlink = 'to.doc', citation_format = 'text', style = 'html')
    # Note links won't show for now due to the following issue
    # https://github.com/cboettig/knitcitations/issues/63

    ## Check all packages (from suggests) needed for the report
    if(is.null(txdb)) load_check('TxDb.Hsapiens.UCSC.hg19.knownGene')
    load_check(c(
        'bumphunter', 'derfinderPlot','ggbio', 'ggplot2', 'grid', 'gridExtra',
        'RColorBrewer', 'mgcv', 'whisker', 'DT', 'sessioninfo'
    ))

    ## Write bibliography information
    bib <- c(
        knitcitations = citation('knitcitations'),
        regionReport = citation('regionReport')[1],
        derfinderPlot = citation('derfinderPlot')[1],
        DT = citation('DT'),
        ggbio = citation('ggbio'),
        ggplot2 = citation('ggplot2'),
        knitr = citation('knitr')[3],
        rmarkdown = citation('rmarkdown'),
        whisker = citation('whisker'),
        bumphunter = citation('bumphunter')[1],
        derfinder = citation('derfinder')[1])
    write.bibtex(bib, file = file.path(outdir, paste0(output, '.bib')))

    ## Save the call
    theCall <- match.call()

    ## Generate report
    ## Perform code within the output directory.
    tmpdir <- getwd()
    with_wd(outdir, {
        file.copy(template, to = paste0(output, '.Rmd'))

        ## Output format
        output_format <- .advanced_argument('output_format',
            'BiocStyle::html_document', ...)
        outputIsHTML <- output_format %in% c('html_document',
            'rmarkdown::html_document',
            'knitrBootstrap::bootstrap_document', 'BiocStyle::html_document')
        if(!outputIsHTML) {
            if(device == 'png') warning("You might want to switch the 'device' argument from 'png' to 'pdf' for better quality plots.")
        }

        ## Check knitrBoostrap version
        knitrBootstrapFlag <- packageVersion('knitrBootstrap') < '1.0.0'
            if(knitrBootstrapFlag & output_format == 'knitrBootstrap::bootstrap_document') {
            ## CRAN version
            tmp <- knit_bootstrap(paste0(output, '.Rmd'), chooser = c('boot',
                'code'), show_code = TRUE)
            res <- file.path(tmpdir, outdir, paste0(output, '.html'))
            unlink(paste0(output, '.md'))
        } else {
            res <- render(paste0(output, '.Rmd'), output_format,
                clean = .advanced_argument('clean', TRUE, ...))
        }
        if(templateNull) file.remove(paste0(output, '.Rmd'))

        ## Open
        if (browse) browseURL(res)
    })

    ## Finish
    return(invisible(res))
}


#' @rdname renderReport
#' @export
templatePvalueDensity <- "
## {{{densityVarName}}}

```{r pval-density-{{{varName}}}, fig.width=10, fig.height=10}
p1{{{varName}}} <- ggplot(regions.df.plot, aes(x={{{varName}}}, colour=seqnames)) +
    geom_line(stat='density') + xlim(0, 1) +
    labs(title='Density of {{{densityVarName}}}') + xlab('{{{densityVarName}}}') +
    scale_colour_discrete(limits=chrs) + theme(legend.title=element_blank())
p1{{{varName}}}
```


```{r 'pval-summary-{{{varName}}}'}
summary(mcols(regions)[['{{{varName}}}']])
```


This is the numerical summary of the distribution of the {{{densityVarName}}}.

```{r pval-tableSummary-{{{varName}}}, results='asis'}
{{{varName}}}table <- lapply(c(1e-04, 0.001, 0.01, 0.025, 0.05, 0.1, 0.2, 0.3, 0.4, 0.5,
    0.6, 0.7, 0.8, 0.9, 1), function(x) {
    data.frame('Cut' = x, 'Count' = sum(mcols(regions)[['{{{varName}}}']] <= x))
})
{{{varName}}}table <- do.call(rbind, {{{varName}}}table)
if(outputIsHTML) {
    kable({{{varName}}}table, format = 'markdown', align = c('c', 'c'))
} else {
    kable({{{varName}}}table)
}
```

This table shows the number of regions with {{{densityVarName}}} less or equal than some commonly used cutoff values.

"

#' @rdname renderReport
#' @export
templateDensity <- "
## {{{densityVarName}}}

```{r density-{{{varName}}}, fig.width=14, fig.height=14, eval=hasSignificant, echo=hasSignificant}
xrange <- range(regions.df.plot[, '{{{varName}}}']) * c(0.95, 1.05)
p3a{{{varName}}} <- ggplot(regions.df.plot[is.finite(regions.df.plot[, '{{{varName}}}']), ], aes(x={{{varName}}}, colour=seqnames)) +
    geom_line(stat='density') + labs(title='Density of {{{densityVarName}}}') +
    xlab('{{{densityVarName}}}') + scale_colour_discrete(limits=chrs) +
    xlim(xrange) + theme(legend.title=element_blank())
p3b{{{varName}}} <- ggplot(regions.df.sig[is.finite(regions.df.sig[, '{{{varName}}}']), ], aes(x={{{varName}}}, colour=seqnames)) +
    geom_line(stat='density') +
    labs(title='Density of {{{densityVarName}}} (significant only)') +
    xlab('{{{densityVarName}}}') + scale_colour_discrete(limits=chrs) +
    xlim(xrange) + theme(legend.title=element_blank())
grid.arrange(p3a{{{varName}}}, p3b{{{varName}}})
```

```{r density-solo-{{{varName}}}, fig.width=10, fig.height=10, eval=!hasSignificant, echo=!hasSignificant}
p3a{{{varName}}} <- ggplot(regions.df.plot[is.finite(regions.df.plot[, '{{{varName}}}']), ], aes(x={{{varName}}}, colour=seqnames)) +
    geom_line(stat='density') + labs(title='Density of {{{densityVarName}}}') +
    xlab('{{{densityVarName}}}') + scale_colour_discrete(limits=chrs) +
    theme(legend.title=element_blank())
p3a{{{varName}}}
```

This plot shows the density of the {{{densityVarName}}} for all regions. `r ifelse(hasSignificant, 'The bottom panel is restricted to significant regions.', '')`

"

#' @rdname renderReport
#' @export
templateManhattan <- "
## Manhattan {{{densityVarName}}}

```{r manhattan-{{{varName}}}, fig.width=10, fig.height=10, message = FALSE}

regions.manhattan <- regions
mcols(regions.manhattan)[['{{{varName}}}']] <- - log(mcols(regions.manhattan)[['{{{varName}}}']], base = 10)
pMan{{{varName}}} <- plotGrandLinear(regions.manhattan, aes(y = {{{varName}}}, colour = seqnames)) + theme(axis.text.x=element_text(angle=-90, hjust=0)) + ylab('-log10 {{{densityVarName}}}')
pMan{{{varName}}}
rm(regions.manhattan)
```

This is a Manhattan plot for the {{{densityVarName}}} for all regions. A single dot is shown for each region, where higher values in the y-axis mean that the {{{densityVarName}}} are closer to zero.

"

#' @rdname renderReport
#' @export
templatePvalueHistogram <- "
## {{{densityVarName}}}

```{r histPval-{{{varName}}}, fig.width=10, fig.height=10}
p1{{{varName}}} <- ggplot(regions.df.plot, aes(x={{{varName}}}, colour=seqnames)) +
    geom_histogram(bins = 50, alpha=.5, position='identity') +
    xlim(c(0, 1.0005)) +
    labs(title='Histogram of {{{densityVarName}}}') +
    xlab('{{{densityVarName}}}') +
    scale_colour_discrete(limits=chrs) + theme(legend.title=element_blank())
p1{{{varName}}}
```

This plot shows the distribution of {{{densityVarName}}} in a histogram. It might be skewed right or left, or flat as shown in the [Wikipedia examples](https://en.wikipedia.org/wiki/Histogram#Examples). The shape depends on the percent of regions that are differentially expressed. For further information on how to interpret a histogram of p-values check [David Robinson's post on this topic](http://varianceexplained.org/statistics/interpreting-pvalue-histogram/).


```{r 'summary-{{{varName}}}'}
summary(mcols(regions)[['{{{varName}}}']])
```


This is the numerical summary of the distribution of the {{{densityVarName}}}.

```{r tableSummary-{{{varName}}}, results='asis'}
{{{varName}}}table <- lapply(c(1e-04, 0.001, 0.01, 0.025, 0.05, 0.1, 0.2, 0.3, 0.4, 0.5,
    0.6, 0.7, 0.8, 0.9, 1), function(x) {
    data.frame('Cut' = x, 'Count' = sum(mcols(regions)[['{{{varName}}}']] <= x))
})
{{{varName}}}table <- do.call(rbind, {{{varName}}}table)
if(outputIsHTML) {
    kable({{{varName}}}table, format = 'markdown', align = c('c', 'c'))
} else {
    kable({{{varName}}}table)
}
```

This table shows the number of regions with {{{densityVarName}}} less or equal than some commonly used cutoff values.

"

#' @rdname renderReport
#' @export
templateHistogram <- "

## {{{densityVarName}}}

```{r histogram-{{{varName}}}, fig.width=14, fig.height=14, eval=hasSignificant, echo=hasSignificant}
xrange <- range(regions.df.plot[, '{{{varName}}}'])
p3a{{{varName}}} <- ggplot(regions.df.plot[is.finite(regions.df.plot[, '{{{varName}}}']), ], aes(x={{{varName}}}, fill=seqnames)) +
    geom_histogram(alpha=.5, position='identity', bins = 50) +
    labs(title='Histogram of {{{densityVarName}}}') +
    xlab('{{{densityVarName}}}') +
    xlim(xrange) + theme(legend.title=element_blank())
p3b{{{varName}}} <- ggplot(regions.df.sig[is.finite(regions.df.sig[, '{{{varName}}}']), ], aes(x={{{varName}}}, fill=seqnames)) +
    geom_histogram(alpha=.5, position='identity', bins = 50) +
    labs(title='Histogram of {{{densityVarName}}} (significant only)') +
    xlab('{{{densityVarName}}}') +
    xlim(xrange) + theme(legend.title=element_blank())
grid.arrange(p3a{{{varName}}}, p3b{{{varName}}})
```

```{r histogram-solo-{{{varName}}}, fig.width=10, fig.height=10, eval=!hasSignificant, echo=!hasSignificant}
p3a{{{varName}}} <- ggplot(regions.df.plot[is.finite(regions.df.plot[, '{{{varName}}}']), ], aes(x={{{varName}}}, fill=seqnames)) +
    geom_histogram(alpha=.5, position='identity', bins = 50) +
    labs(title='Histogram of {{{densityVarName}}}') +
    xlab('{{{densityVarName}}}') +
    theme(legend.title=element_blank())
p3a{{{varName}}}
```

This plot shows the distribution of {{{densityVarName}}} in a histogram for all regions. `r ifelse(hasSignificant, 'The bottom panel is restricted to significant regions.', '')`

"
